### PR TITLE
DAOS-12063 object: reset kds and recxs for each shard

### PR DESCRIPTION
**Pull request by wangdi1**
During EC object enumeration, kds and recxs over each sub anchor may need reallocate, if the number is changed.

set rd_lvl to rank, in the case the test runs with more ranks and fewer nodes.

Required-githooks: true

Signed-off-by: Di Wang <di.wang@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/suite/dfs_unit_test.c [hunk 1]
```diff
@@ -1906,6 +1908,13 @@ dfs_test_readdir(void **state)
 	assert_int_equal(rc, 0);
 }
 
+static void
+dfs_test_readdir(void **state)
+{
+	dfs_test_readdir_internal(state, OC_SX);
+	dfs_test_readdir_internal(state, OC_EC_2P2GX);
```
Comment (mchaarawi) [inline src/tests/suite/dfs_unit_test.c:1915]: we should add a check if this oclass is runnable on current system (i.e. should skip if we have less than 4 engines or whatever the rd level is for this oclass to work)

File: src/tests/suite/dfs_unit_test.c [hunk 2]
```diff
@@ -2212,15 +2227,22 @@ dfs_setup(void **state)
...
 		if (use_dtx)
 			print_message("Running DFS Serial tests with DTX enabled\n");
 		else
 			print_message("Running DFS Serial tests with DTX disabled\n");
 
-		rc = dfs_cont_create(arg->pool.poh, &co_uuid, NULL, &co_hdl, &dfs_mt);
+		attr.da_props = daos_prop_alloc(1);
+		assert_non_null(attr.da_props);
+		attr.da_props->dpp_entries[0].dpe_type = DAOS_PROP_CO_REDUN_LVL;
+		attr.da_props->dpp_entries[0].dpe_val = DAOS_PROP_CO_REDUN_RANK;
```
Comment (mchaarawi) [inline src/tests/suite/dfs_unit_test.c:2242]: that is not the default? i would rather we use the default with this test TBH as that what most users would use.

Comment (wangdi1) [reply src/tests/suite/dfs_unit_test.c:2242]: default is node, not rank. Actually this is mostly for running a test with many engines over a single node. Anyway, this is probably out of scope of this ticket.

File: src/tests/suite/dfs_unit_test.c (hunk 1 above)
Comment (wangdi1) [reply src/tests/suite/dfs_unit_test.c:1915]: ah, sure

**Review by jolivier23 — APPROVED** [review]

File: src/object/cli_obj.c [hunk 3]
```diff
@@ -5598,8 +5601,11 @@ shard_anchors_check_alloc_bufs(struct obj_auxi_args *obj_auxi,
 			req_tgts->ort_shard_tgts[i].st_rank = DAOS_TGT_IGNORE;
 		}
 
-		if (sub_anchor->ssa_recxs != NULL)
-			continue;
+		if (sub_anchor->ssa_recxs != NULL) {
+			if (sub_anchors->sa_nr == nr)
+				continue;
+			D_FREE(sub_anchor->ssa_recxs);
```
Comment (jolivier23) [inline src/object/cli_obj.c:5607]: Do you still need to address the issue of init/fini to deallocate the sub_anchors when the iteration terminates early?

**Review by mchaarawi — APPROVED** [review]

File: src/object/cli_obj.c (hunk 3 above)
Comment (wangdi1) [reply src/object/cli_obj.c:5607]: oh, I will resolve that in a different patch.

**Review by liuxuezhao — APPROVED** [review]

**Comment by mchaarawi** [comment]
failure is known issue? can we land this now?

**Pull request merged**